junit-platform-launcher/src/main/java/org/junit/platform/launcher/TagFilter.java

```diff
@@ -126,15 +127,30 @@ public static PostDiscoveryFilter excludeTags(List<String> tagExpressions) throw
 		return includeMatching(tagExpressions, Stream::noneMatch);
 	}
 
+	private static String inclusionReasonTagContains(Set<TestTag> tags) {
+		return String.format("Test included Because it contains tags: [%s]", formatToString(tags));
+	}
+
+	private static String exclusionReasonTagContains(Set<TestTag> tags) {
+		return String.format("Test excluded Because it contains tags: [%s]", formatToString(tags));
+
+	}
+
+	private static String formatToString(Set<TestTag> testTags) {
+		return testTags.stream().map(TestTag::getName).collect(Collectors.joining(","));
+	}
+
 	private static PostDiscoveryFilter includeMatching(List<String> tagExpressions,
 			BiPredicate<Stream<TagExpression>, Predicate<TagExpression>> matcher) {
 
 		Preconditions.notEmpty(tagExpressions, "list of tag expressions must not be null or empty");
 		List<TagExpression> parsedTagExpressions = parseAll(tagExpressions);
 		return descriptor -> {
 			Set<TestTag> tags = descriptor.getTags();
-			return FilterResult.includedIf(
-				matcher.test(parsedTagExpressions.stream(), expression -> expression.evaluate(tags)));
+			boolean included = matcher.test(parsedTagExpressions.stream(), expression -> expression.evaluate(tags));
+
+			return FilterResult.includedIf(included, () -> inclusionReasonTagContains(tags),
+				() -> exclusionReasonTagContains(tags));
```

Comment:
The main issue I see here is that the provided reason will not always be true, if I'm not mistaken.

Say that we have a class with `tag2` and we only want to include tests with `tag1`. When the filter will be applied to this class, the `included` boolean here will be `false`. It will result in the following exclusion reason: `"Test excluded because it contains tags: [tag2]"`. But shouldn't it actually be `"Test excluded because it does *not* contains tags: [tag1]"`?

The issue is that here, you don't have the original intention to *include* or *exclude* the provided tags, since it's hidden inside the `matcher`...

Reply:
On another note, do we actually need an inclusion reason?

Reply:
@juliette-derancourt I thought of this but was not clear if we require different exclusion reason because of their tags. Although  ``` "Test excluded because it does *not* contains tags: [tag1]"?``` this seems more accurate. I added inclusion reason to maintain symmetry, isn't it relevant to add inclusion reason  as well,

Reply:
> I thought of this but was not clear if we require different exclusion reason because of their tags. Although "Test excluded because it does *not* contains tags: [tag1]"? this seems more accurate.

I'm not saying we should use this wording or even do this, I'm just saying that the reason you're printing now is false in the case of a "tag inclusion" :)
We actually need to decide if we want to log something if we're in that case (i.e. if some tests are not executed because they don't have the "included tag" on them) or if we're just sticking to the "excluded tag" case. (Don't know if this makes sense outside of my head ahah)

> I added inclusion reason to maintain symmetry, isn't it relevant to add inclusion reason as well?

Well you're not using the inclusion reason in your current implementation, only the exclusion one (see in [here](https://github.com/junit-team/junit5/pull/2043/files#diff-1b699f4d83a93825dbcb40dc72f85278R97)). But the symmetry argument makes sense I guess

Reply:
Here What we can do is, we can check whether this method is called from includeTags or excludeTags method. accordingly, we need to populate the message, 

Say that we have a class with tag2 and we only want to include tests with tag1, meaning caller function is **includeTags()**, When the filter will be applied to this class, the included boolean here will be false. Hence we need to give exclusion reason for method, which can be 
like '**Test excluded because the method does not contain any of the tags: [tag1]**'

I believe this makes much more sense since one has to know the methods and reason for which tests have not executed. if we agree then I can push the changes along with other changes.

Reply:
@ptlanu22 Sorry to keep you waiting, yes let’s go with the "does not contain" reason if the test is not explicitely included.

I guess you could remove the matcher parameter and make two methods to separate the include/exclude cases ;)

Reply:
We are actually handling tag expressions at this point. Thus, we want to state something along the lines of:


```
The following containers and tests were excluded since their tags do not match the tag expression '<expression>': <containers and tests>
```

Reply:
Well here we are giving the exclusion reason for one test/container, the actual reduce + logging happens in the launcher 👇
Therefore, the wording of the reason does not really matter, right?

Reply:
Well here we are giving the exclusion reason for one test/container, the actual reduce + logging happens in the launcher 👇
Therefore, the wording of the reason does not really matter, right?

Reply:
Sorry for getting this out of context.

Yes, the aggregation occurs in the launcher.

In any case, the wording does matter. A test does not "contain" tags. Rather, a test or container can be "assigned" tags, and the test or container is then included or excluded based on whether the assigned tags match the supplied tag expression.

Hope that's clearer.

Reply:
Yes of course, I’m not saying it _doesn’t_ matter (I know I just said that 😂), but I meant that it will not be the actual thing that will be in the log, it’s just a key for aggregating the reasons :)
But yeah you’re right, contain is not the right word!

Reply:
@juliette-derancourt / @sbrannen , Here I believe we have to compromise on inclusion/exclusion reasons. so instead of tags, I changed the message to expression, that is still not perfect, but better than a message with Tags.